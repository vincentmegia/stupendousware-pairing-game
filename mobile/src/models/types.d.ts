export type CardModel = {
  id: number
  value: string
  matched: boolean
  selected: boolean
  reset: boolean
}
