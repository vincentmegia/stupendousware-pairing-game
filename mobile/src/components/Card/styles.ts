import {StyleSheet} from 'react-native'
export default StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.09,
    shadowRadius: 4,
    elevation: 5,
  },
  cardFront: {
    justifyContent: 'center',
    alignItems: 'center',
    minHeight: 130,
    maxHeight: 140,
    minWidth: 100,
    maxWidth: 110,
    backgroundColor: 'white',
    borderRadius: 20,
    position: 'absolute',
  },
  cardBack: {
    justifyContent: 'center',
    alignItems: 'center',
    minHeight: 130,
    maxHeight: 140,
    minWidth: 100,
    maxWidth: 110,
    backgroundColor: 'white',
    borderRadius: 20,
    position: 'absolute',
    backfaceVisibility: 'hidden',
  },
  textBack: {
    backgroundColor: 'white',
    color: 'white',
    display: 'none',
    backfaceVisibility: 'hidden',
  },
  textFront: {
    backfaceVisibility: 'hidden',
    position: 'absolute',
    fontWeight: '900',
  },
})
