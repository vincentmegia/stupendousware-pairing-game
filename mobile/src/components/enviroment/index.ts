import {createContext} from 'react'

export const configuration = {environment: 'production'}
export const EnvironmentContext = createContext(configuration)
