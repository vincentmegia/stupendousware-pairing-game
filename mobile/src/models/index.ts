import {CardModel} from './types'

export function createCard(id: number, value: string): CardModel {
  return {
    id,
    value,
    matched: false,
    selected: false,
    reset: false,
  }
}

export const findCard = (cards: Array<CardModel[]>, id: number): CardModel => {
  for (const columns of cards) {
    for (const column of columns) {
      if (column.id === id) {
        return column
      }
    }
  }
  return createCard(-1, '-1')
}

export const updateCard = (cards: Array<CardModel[]>, card: CardModel) => {
  for (let row = 0; row < cards.length; row++) {
    for (let column = 0; column < cards[row].length; column++) {
      if (cards[row][column].id === card.id) {
        cards[row][column] = card
        break
      }
    }
  }
}
