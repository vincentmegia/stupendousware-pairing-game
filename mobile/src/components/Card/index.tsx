import React, {memo, useCallback, useEffect, useRef, useState} from 'react'
import {Animated, Pressable} from 'react-native'
import {CardModel} from '@models/types'
import Debug from '../debug'
import styles from './styles'

interface CardProps {
  card: CardModel
  onSelected: (card: CardModel) => void
}
const Card = (props: CardProps) => {
  const {card, onSelected} = props
  const flipAnimation = useRef(new Animated.Value(0)).current
  const flipRef = useRef(false)
  const [selected, setSelected] = useState(card.selected)

  useEffect(() => {
    if (card.reset) {
      flipToBack()
      setSelected(card.selected)
      card.reset = false
    }
  }, [card.reset])

  /**
   * Flip to front animation
   */
  const flipToFrontStyle = {
    transform: [
      {
        rotateY: flipAnimation.interpolate({
          inputRange: [0, 180],
          outputRange: ['0deg', '180deg'],
        }),
      },
    ],
  }

  /**
   * Flip to back animation
   */
  const flipToBackStyle = {
    transform: [
      {
        rotateY: flipAnimation.interpolate({
          inputRange: [0, 180],
          outputRange: ['180deg', '360deg'],
        }),
      },
    ],
  }

  const flipToFront = useCallback(() => {
    flipRef.current = true
    Animated.timing(flipAnimation, {
      toValue: 180,
      duration: 300,
      useNativeDriver: true,
    }).start()
  }, [])

  const flipToBack = useCallback(() => {
    if (card.matched) return
    flipRef.current = false
    Animated.timing(flipAnimation, {
      toValue: 0,
      duration: 300,
      useNativeDriver: true,
    }).start()
  }, [])

  const onPress = useCallback(() => {
    if (flipRef.current) {
      flipToBack()
    } else {
      flipToFront()
    }
    setSelected(!selected)
    onSelected({...card, selected: !card.selected})
  }, [])

  return (
    <>
      <Pressable style={styles.container} onPress={onPress} disabled={selected}>
        <Animated.View style={{...styles.cardFront, ...flipToBackStyle}}>
          <Animated.Text style={{...styles.textFront, ...flipToBackStyle}}>
            {card.value}
          </Animated.Text>
          <Debug />
        </Animated.View>
        <Animated.View style={{...styles.cardBack, ...flipToFrontStyle}}>
          <Animated.Text style={{...styles.textBack, ...flipToFrontStyle}}>
            {card.value}
          </Animated.Text>
          <Debug />
        </Animated.View>
      </Pressable>
    </>
  )
}

export default memo(Card)
