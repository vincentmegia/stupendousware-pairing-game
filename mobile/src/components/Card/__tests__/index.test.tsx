import React from 'react'
import Card from '..'
import {render} from '@testing-library/react-native'
import {createCard} from '../../../models'
import {configuration, EnvironmentContext} from '../../enviroment'

describe('Card component tests', () => {
  it('it matches snapshot', () => {
    const {toJSON} = render(
      <EnvironmentContext.Provider value={{environment: 'test'}}>
        <Card card={createCard(1, '')} onSelected={jest.fn()} />
      </EnvironmentContext.Provider>,
    )

    expect(toJSON()).toMatchSnapshot()
  })
})
