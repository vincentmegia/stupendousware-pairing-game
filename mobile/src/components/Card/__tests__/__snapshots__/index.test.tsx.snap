// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component tests it matches snapshot 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "elevation": 5,
      "justifyContent": "center",
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.09,
      "shadowRadius": 4,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 20,
        "justifyContent": "center",
        "maxHeight": 140,
        "maxWidth": 110,
        "minHeight": 130,
        "minWidth": 100,
        "position": "absolute",
        "transform": Array [
          Object {
            "rotateY": "180deg",
          },
        ],
      }
    }
  >
    <Text
      collapsable={false}
      style={
        Object {
          "backfaceVisibility": "hidden",
          "fontWeight": "900",
          "position": "absolute",
          "transform": Array [
            Object {
              "rotateY": "180deg",
            },
          ],
        }
      }
    />
    <Text
      collapsable={false}
      style={
        Object {
          "color": "grey",
          "fontSize": 10,
          "marginTop": 80,
        }
      }
    >
      0
    </Text>
  </View>
  <View
    collapsable={false}
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "white",
        "borderRadius": 20,
        "justifyContent": "center",
        "maxHeight": 140,
        "maxWidth": 110,
        "minHeight": 130,
        "minWidth": 100,
        "position": "absolute",
        "transform": Array [
          Object {
            "rotateY": "0deg",
          },
        ],
      }
    }
  >
    <Text
      collapsable={false}
      style={
        Object {
          "backfaceVisibility": "hidden",
          "backgroundColor": "white",
          "color": "white",
          "display": "none",
          "transform": Array [
            Object {
              "rotateY": "0deg",
            },
          ],
        }
      }
    />
    <Text
      collapsable={false}
      style={
        Object {
          "color": "grey",
          "fontSize": 10,
          "marginTop": 80,
        }
      }
    >
      0
    </Text>
  </View>
</View>
`;
